File: examples/koog-java-api-example/src/main/java/org/example/koog/java/agents/KoogAgentService.java
package org.example.koog.java.agents;

import ai.koog.agents.core.agent.AIAgent;
import ai.koog.agents.core.agent.AIAgentState;
import ai.koog.agents.core.agent.config.AIAgentConfig;
import ai.koog.agents.core.agent.context.AIAgentFunctionalContext;
import ai.koog.agents.core.tools.Tool;
import ai.koog.agents.core.tools.ToolRegistry;
import ai.koog.agents.features.eventHandler.feature.EventHandler;
import ai.koog.prompt.dsl.Prompt;
import ai.koog.prompt.executor.clients.anthropic.AnthropicModels;
import ai.koog.prompt.executor.clients.openai.OpenAIModels;
import ai.koog.prompt.executor.llms.MultiLLMPromptExecutor;
import io.opentelemetry.sdk.trace.export.SpanExporter;
import org.example.koog.java.structs.OrderSupportRequest;
import org.example.koog.java.structs.OrderUpdateSummary;
import org.example.koog.java.tools.UserTools;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


/**
 * Java conversion of KoogAgentService.
 * <p>
 * Notes:
 * - Kotlin suspend interop: the original methods were suspend. This Java version runs the agent
 * logic asynchronously on a fixed thread pool and returns immediately where applicable.
 * - AIAgent construction uses a Kotlin-first DSL in the original code. The creation block is marked
 * with TODOs so you can connect it to the Java-accessible API (constructor or builder).
 * - createPostgresStorage() is a Kotlin top-level function. Replace the "Kt" class name below with
 * the actual one generated from the Kotlin file name that defines createPostgresStorage().
 */
@Service
public class KoogAgentService {

    private final MultiLLMPromptExecutor promptExecutor;
    private final List<SpanExporter> spanExporters;

    // userId -> set of agentIds (Boolean.TRUE to emulate Kotlin's Unit placeholder)
    private final ConcurrentMap<String, ConcurrentMap<String, Boolean>> agentIdsByUser = new ConcurrentHashMap<>();
    private final ConcurrentMap<String, AIAgent<String, OrderUpdateSummary>> agentsById = new ConcurrentHashMap<>();

    // Lazily initialized Postgres storage created by Kotlin top-level function.
    // Replace the type with the concrete storage interface/class used by Persistence.storage.
    private volatile Object postgresStorage;

    // Fixed thread pool similar to Kotlin newFixedThreadPoolContext
    private final ExecutorService parallelExecutor = Executors.newFixedThreadPool(10, r -> {
        Thread t = new Thread(r);
        t.setName("agents");
        t.setDaemon(true);
        return t;
    });

    public KoogAgentService(
        MultiLLMPromptExecutor promptExecutor,
        List<SpanExporter> spanExporters
    ) {
        this.promptExecutor = Objects.requireNonNull(promptExecutor, "promptExecutor");
        this.spanExporters = Objects.requireNonNull(spanExporters, "spanExporters");
    }

    private AIAgent<String, OrderUpdateSummary> createAgent(String userId) {
        UserTools userTools = new UserTools(userId);

        List<Tool<?, ?>> readingTools = List.of(
            userTools.getTool("readUserAccount"),
            userTools.getTool("readUserOrders")
        );

        var agent = AIAgent.builder()
            .promptExecutor(promptExecutor)
            .agentConfig(
                AIAgentConfig.builder()
                    .model(OpenAIModels.Chat.GPT4_1)
                    .prompt(Prompt.builder("id")
                        .system("system")
                        .user("user")
                        .assistant("assistant")
                        .user("user")
                        .assistant("assistant")
                        .toolCall("id-1", "tool-1", "args-1")
                        .toolResult("id-1", "tool-1", "result-1")
                        .toolCall("id-2", "tool-2", "args--2")
                        .toolResult("id-2", "tool-2", "result-2")
                        .build()
                    )
                    .maxAgentIterations(100)
                    .strategyExecutorService(Executors.newSingleThreadExecutor()) // TODO: use better executor service
                    .llmRequestExecutorService(Executors.newSingleThreadExecutor()) // TODO: use better executor service
                    .build()
            )
            .functionalStrategy((AIAgentFunctionalContext context, String userInput) -> {
                var orderRequest = context
                    .subtask("Identify the user's problem. Input: " + userInput + ".")
                    .withInput(userInput)
                    .withOutput(OrderSupportRequest.class)
                    .withTools(readingTools)
                    .useLLM(OpenAIModels.Chat.GPT4o)
                    .run();

                if (orderRequest.isResolved()) {
                    return orderRequest.emptyUpdate();
                }

                var update = context
                    .subtask("Solve the identified user's problem")
                    .withInput(orderRequest)
                    .withOutput(OrderUpdateSummary.class)
                    .withTools(userTools.asTools())
                    .useLLM(AnthropicModels.Sonnet_4)
                    .run();

                while (true) {
                    var verification = context
                        .subtask("Verify if the initial user's problem was solved")
                        .withInput(update)
                        .withVerification()
                        .withTools(readingTools)
                        .useLLM(OpenAIModels.Chat.O3)
                        .run();

                    if (verification.getSuccessful()) {
                        return update;
                    } else {
                        update = context
                            .subtask("You must resolve the following issues: " + verification.getFeedback())
                            .withInput(orderRequest)
                            .withOutput(OrderUpdateSummary.class)
                            .useLLM(AnthropicModels.Haiku_4_5)
                            .withTools(userTools.asTools())
                            .run();
                    }
                }

            })
            .toolRegistry(
                ToolRegistry.builder()
                    .tools(userTools)
                    .build()
            )
//                TODO: migrate to graphs
//                .install(Persistence.Feature, config -> {
//                    config.setEnableAutomaticPersistence(true);
//                    config.setRollbackToolRegistry(
//                            RollbackToolRegistry.builder()
//                                    .registerRollback(new MyTool(), new MyTool())
//                                    .registerRollbacks(userTools, rollbackTools)
//                                    .build()
//                    );
//                })
//                .install(OpenTelemetry.Feature, config -> {
//                    config.setVerbose(true);
//                    config.addSpanExporter(null);
//                })
            .install(EventHandler.Feature, config -> {
                config.onToolCallStarting(ctx -> {
                    System.out.println("tool called: " + ctx.getToolName());
                });
                config.onAgentClosing(ctx -> {
                    System.out.println("agent finishing: " + ctx.getAgentId());
                });
            })
            .build();

        agentIdsByUser
            .computeIfAbsent(userId, k -> new ConcurrentHashMap<>())
            .put(agent.getId(), Boolean.TRUE);

        agentsById.put(agent.getId(), agent);
        return agent;
    }

    /**
     * Starts an agent asynchronously and returns its id immediately.
     */
    public String launchSupportAgent(String userId, String question) {
        AIAgent<String, OrderUpdateSummary> agent = createAgent(userId);
        parallelExecutor.submit(() -> {
            try {
                // If agent.run is suspend in Kotlin, wrap it with runBlocking in a small Kotlin helper,
                // or expose a Java-friendly run method.
                agent.run(question);
            } catch (Throwable t) {
                // Log the failure as appropriate for your project
            }
        });
        return agent.getId();
    }

    public AIAgentState<OrderUpdateSummary> getState(String agentId) throws Exception {
        AIAgent<String, OrderUpdateSummary> agent = agentsById.get(agentId);
        if (agent == null) {
            throw new Exception("Agent with id = `" + agentId + "` not found");
        }
        // If getState is suspend, expose a Java-friendly wrapper or call via runBlocking.
        return agent.getState();
    }

    public List<String> getAgentIds(String userId) {
        ConcurrentMap<String, Boolean> ids = agentIdsByUser.get(userId);
        if (ids == null) {
            return Collections.emptyList();
        }
        return new ArrayList<>(ids.keySet());
    }
}

File: examples/devoxx-belgium-2025/src/main/kotlin/ai/koog/spring/sandwich/agents/KoogAgentService.kt
package ai.koog.spring.sandwich.agents

import ai.koog.agents.core.agent.AIAgent
import ai.koog.agents.core.agent.AIAgentState
import ai.koog.agents.core.tools.ToolRegistry
import ai.koog.agents.core.tools.reflect.tools
import ai.koog.agents.features.opentelemetry.feature.OpenTelemetry
import ai.koog.agents.snapshot.feature.AgentCheckpointData
import ai.koog.agents.snapshot.feature.Persistence
import ai.koog.agents.snapshot.feature.RollbackToolRegistry
import ai.koog.agents.snapshot.feature.withPersistence
import ai.koog.prompt.executor.clients.openai.OpenAIModels
import ai.koog.prompt.executor.llms.MultiLLMPromptExecutor
import ai.koog.prompt.executor.model.PromptExecutor
import ai.koog.spring.sandwich.checkpoints.createPostgresStorage
import ai.koog.spring.sandwich.structs.OrderUpdateSummary
import ai.koog.spring.sandwich.tools.CommunicationTools
import ai.koog.spring.sandwich.tools.OrderTools
import ai.koog.spring.sandwich.tools.RollbackTools
import ai.koog.spring.sandwich.tools.UserTools
import io.ktor.util.collections.*
import io.opentelemetry.sdk.trace.export.SpanExporter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.launch
import kotlinx.coroutines.newFixedThreadPoolContext
import kotlin.time.Instant
import kotlinx.serialization.Serializable
import org.springframework.boot.info.BuildProperties
import org.springframework.stereotype.Service

@Service
class KoogAgentService(
    private val promptExecutor: PromptExecutor,
    private val spanExporters: List<SpanExporter>,
    private val buildProps: BuildProperties,
) {
    private val agentIdsByUser = ConcurrentMap<String, ConcurrentMap<String, Unit>>()
    private val agentsById = ConcurrentMap<String, AIAgent<String, OrderUpdateSummary>>()

    private val postgresStorage by lazy { createPostgresStorage() }

    private fun createAgent(userId: String): AIAgent<String, OrderUpdateSummary> {
        val userTools = UserTools(userId)

        val rollbackTools = RollbackTools(userId)

        val agent = AIAgent(
            promptExecutor = promptExecutor,
            llmModel = OpenAIModels.Chat.GPT5Nano,
            toolRegistry = ToolRegistry {
                tools(CommunicationTools)
                tools(OrderTools)
                tools(userTools)
            },
            strategy = interactiveSupportStrategy(userTools),
        ) {
            install(OpenTelemetry) {
                setServiceInfo(serviceName = buildProps.name!!, serviceVersion = buildProps.version!!)
                setVerbose(true)
                spanExporters.forEach {
                    addSpanExporter(it)
                }
            }

            install(Persistence) {
                storage = postgresStorage

                enableAutomaticPersistence = true

                rollbackToolRegistry = RollbackToolRegistry {
                    registerRollback(userTools::issueRefund, rollbackTools::undoRefund)
                    registerRollback(userTools::makeAnotherOrder, rollbackTools::undoAnotherOrder)
                    registerRollback(OrderTools::contactCarrier, rollbackTools::notifyCarrierAboutCancellation)
                    registerRollback(OrderTools::updateAddress, rollbackTools::rollbackAddress)
                }
            }
        }

        agentIdsByUser.getOrPut(userId) { ConcurrentMap() }[agent.id] = Unit
        agentsById[agent.id] = agent
        return agent
    }

    @OptIn(DelicateCoroutinesApi::class)
    private val parallelScope = CoroutineScope(newFixedThreadPoolContext(nThreads = 10, name = "agents"))

    suspend fun launchSupportAgent(userId: String, question: String): String {
        val agent = createAgent(userId)
        parallelScope.launch { agent.run(question, null) }
        return agent.id
    }

    suspend fun getState(agentId: String): State<OrderUpdateSummary> {
        val agent = agentsById[agentId] ?: throw Exception("Agent with id = `$agentId` not found")
        return agent.getState()
    }

    @Serializable
    data class CheckpointInfo(
        val checkpointId: String,
        val createdAt: Instant,
        val nodeId: String
    ) {
        companion object {
            fun fromCheckpoint(checkpoint: AgentCheckpointData): CheckpointInfo = CheckpointInfo(
                checkpoint.checkpointId, checkpoint.createdAt, checkpoint.nodeId
            )
        }
    }

    suspend fun getCheckpoints(agentId: String): List<CheckpointInfo> =
        postgresStorage.getCheckpoints(agentId).map(CheckpointInfo::fromCheckpoint)

    suspend fun getAgentIds(userId: String): List<String> =
        agentIdsByUser[userId]?.keys?.toList() ?: emptyList()

    suspend fun rollback(agentId: String, checkpointId: String) {
        val agent = agentsById[agentId] ?: throw Exception("Agent with id = `$agentId` not found")
        val session = agent.getSession() ?: throw Exception("Agent session not available")
        session.withPersistence { ctx ->
            rollbackToCheckpoint(checkpointId, ctx)
        }
    }
}
